feat: Insert history and versions endpoint

n-dusan opened the pull request:
Closes #17, #24

This PR is the first-pass implementation for inserting data from an RDF git repo into the database. We refer to the type of data in RDF that we work with as _changes_, _versions_ or _historical_ information about documents or collections. We refer to documents as legal material in a stele. Documents can get organized/categorized into (sub)directories, which we call collections. In addition to tracking changes to documents, we also track changes to collections of documents.

#### Rationale

The end goal is to implement an API that lists all the significant dates (that we refer to also as versions) that affected legal documents or collections for a stele across history. 

#### Notes

To support the end-goal, this pr's adds: 
1) database access and models the tables necessary for the versions API; 
2) load/parse change information from a stele RDF data repository. 

For database access we used [sqlx](https://docs.rs/sqlx/latest/sqlx/), for parsing/working with the RDF, we used [sophia](https://docs.rs/sophia/latest/sophia).

#### Overview

- Added initial database SQL tables necessary for implementing the historical versions API. 
- Added `db` module that's supposed to be generic over database connections. Currently supports connecting and inserting changes to SQLite database with `sqlx`. There's also Postgres, because we left open the possibility to easily extend into Postgres. Postgres support is still TBD. If we do decide that only SQLite is the way to go, we'll remove the Postgres code in a separate pr.
- The `changes` module parses the changes into a graph and inserts the data into db. The `rdf` module has utility methods supposed to reduce boilerplate and make working with the graph easier. 
- `migrations` directory has the initial database schema/tables. 
- `models` contains sqlx structs that map to database tables.
- `statements` module has SQL `inserts` and `queries` modules for SQL statements.

Introduce the /_api/versions endpoint. Idea being we query the database and, depending on the url request, return either all document or collection changes. The changes are mapped onto a `response::Versions` json response.

#### To implement in a separate pr

- Add integration testing framework that has a small RDF repository fixture. Then we call in the insert history command and assert expected states.
- I'm thinking of potentially adding traits that tie database methods to the model in `models`.  Something like a `PublicationManager`, for example. Using traits could make the `inserts` and `queries` more legible. 

We'll need to update the insert changes logic again, so I vote to defer [this](https://github.com/openlawlibrary/stelae/issues/35) and [this](https://github.com/openlawlibrary/stelae/issues/34) issue at that point.

tombh reviewed (changes requested):
This looks really good! I can't see any significant issues!

The biggest thing to me is actually the commit history. I think it'd be useful to start getting into the habit of `git commit --amend` and `git rebase --interactive` so that as much as possible the commits tell a narrative-like story. I guess you'll be doing a squash rebase to merge this PR? Of course that's fine, but I think it'd be worth bearing in mind how the commit history reads as you're actually in the process of developing the PR. I'm guessing that with the help of `--amend` and `rebase --interactive` this PR could probably be represented in something like ~5 commits? Well-constructed commits are in themselves documentation, think of how `git blame` can add context to future developers (including yourself!) when doing refactors or bug fixing. And of course expressive, story-telling commits can add another medium of context and communication to help the reviewer.

With such habits it can also be good to try to make sure that every commit passes CI. This adds a layer of confidence to each commit, expressing to future developers that reliability of the given code changes. It can also help with merge conflicts, especially those arising from rebased merges, where each commit's conflicts can be more reliably resolved by running the test suite at each step of the rebase.

tombh commented:
Also, would you be interested in a commit lint? It's really easy to add to CI, add a `commitlintrc.yaml` following https://commitlint.js.org/reference/configuration.html (Eg https://github.com/openlawlibrary/pygls/blob/main/commitlintrc.yaml) and the Github Action is then a oneliner: `npx commitlint --from origin/main --to HEAD --verbose --config commitlintrc.yaml`

n-dusan commented on src/db/models/publication.rs:
```diff
@@ -0,0 +1,40 @@
+    #[allow(clippy::unwrap_in_result, clippy::unwrap_used)]
+    fn from_row(row: &AnyRow) -> Result<Self, sqlx::Error> {
+        Ok(Self {
+            name: row.try_get("name").unwrap(),
```
Was not sure `sqlx` integrates with anyhow this well, but it worked! Thank you

n-dusan commented on src/history/rdf/graph.rs:
```diff
@@ -0,0 +1,217 @@
+    /// Errors if the object is not an RDF literal.
+    pub fn literal_from_triple_matching(
+        &self,
+        subject: Option<&SimpleTerm>,
```
That suggestion sounds reasonable, but I believe it would be more readable right now without struct fields.  We pass in these args differently when inserting history, so in that case we'd need to reinitialize the struct fields often.

n-dusan commented on src/db/statements/queries.rs:
```diff
@@ -0,0 +1,284 @@
+        DatabaseKind::Sqlite => {
+            let mut connection = conn.pool.acquire().await?;
+
+            let statement = format!("
```
Did not see it, that looks awesome. I've got a second pr ready for review hopefully soon, so I'd defer adding separate sql files later. Created an [issue](https://github.com/openlawlibrary/stelae/issues/35).

n-dusan commented on src/server/publish.rs:
```diff
@@ -156,24 +176,35 @@ pub async fn serve_archive(
+                "error: could not connect to database. Confirm that DATABASE_URL env var is set correctly."
+            );
+            tracing::error!("Error: {:?}", err);
+            process::exit(1);
```
Thanks, that makes a lot of sense. I've created an [issue](https://github.com/openlawlibrary/stelae/issues/36)for it for now.

n-dusan commented on src/history/rdf/graph.rs:
```diff
@@ -0,0 +1,217 @@
+        loop {
+            let el_uri = format!("http://www.w3.org/1999/02/22-rdf-syntax-ns#_{i}");
+            let elem_iri = SimpleTerm::Iri(IriRef::new_unchecked(MownStr::from_str(&el_uri)));
+            if self
```
Thanks, changed it

n-dusan commented on src/history/rdf/graph.rs:
```diff
@@ -0,0 +1,217 @@
+    pub fn items(&self) -> anyhow::Result<Vec<SimpleTerm>> {
+        let container = &self.uri;
+        let mut i = 1_u32;
+        let mut l_ = vec![];
```
Changed

n-dusan commented:
@tombh Thank you, this has been insightful! Agree about the commit history, so I've brought the PR down from 98 to 7 commits. 

Also a commit linter would be amazing. But I'd say we add that in another pr.

I don't think we need to address all of these comments right now, since I have another PR ready for review hopefully soon. I've created a couple of issues [[1](https://github.com/openlawlibrary/stelae/issues/34), [2](https://github.com/openlawlibrary/stelae/issues/35), [3](https://github.com/openlawlibrary/stelae/issues/36)].

tombh commented on src/db/models/publication.rs:
```diff
@@ -0,0 +1,39 @@
+            stele: row.try_get("stele")?,
+            revoked: row.try_get("revoked")?,
+            last_valid_publication_name: row.try_get("last_valid_publication_name").ok(),
+            last_valid_version: row.try_get("last_valid_version").ok(),
```
The `.ok()`s can be `?` as well right?

n-dusan replied:
Looks like sqlx can't map `?` to Option by default. So can't be `?` right now :-(

tombh commented on src/db/models/publication_version.rs:
```diff
@@ -0,0 +1,26 @@
+            version: row.try_get("version")?,
+            publication: row.try_get("publication")?,
+            stele: row.try_get("stele")?,
+            build_reason: row.try_get("build_reason").ok(),
```
`.ok()` to `?`

tombh commented on src/db/statements/queries.rs:
```diff
@@ -0,0 +1,284 @@
+                .bind(name)
+                .fetch_one(&mut *connection)
+                .await
+                .ok()
```
`.ok()` to `?`

tombh commented on src/db/statements/queries.rs:
```diff
@@ -0,0 +1,284 @@
+                .bind(stele)
+                .fetch_one(&mut *connection)
+                .await
+                .ok()
```
`.ok()` to `?`

tombh reviewed (changes requested):
Just a couple of `ok()`s to look at. Otherwise it looks good!

n-dusan commented:
Thanks, for the rebase pr! I've now rebased your #40 into this PR and addressed comments from the other PR and most of the `.ok`'s are now `?`. Had to keep some of them, since in some cases we don't really want to early exit the function and propagate the error, so None was expected.
